### DISCUSSION TITLE
@OneToMany audit — is it possible to findChanges for child entity fields within the parent entity?

### DISCUSSION THREAD

**Discussion by dmitry-weirdo** [javers/javers#1120] [Q&A]
Unfortunately, I cannot find a documentation not on `@Value` classes (`Address` in the tests), but for `@Entity` classes in the collection of the parent `@Entity` class.

Imagine we have a typical unidirectional `@OneToMany` entities connection:
```java
@javax.persistence.Entity
public class ParentEntity {
// ...

    @JoinColumn(name = "parent_entity_id")
    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<ChildEntity> childEntities;
}
```

When I now save parent entity changes (one transaction with cascade changes in the collection `childEntities` fields), the audit is saved to the database.

However, this `findChanges` query only returns the parent entity fields. 
```java
QueryBuilder queryBuilder = QueryBuilder.byInstanceId(parentEntityId, ParentEntity.class);

javers.findChanges(queryBuilder.build()).groupByCommit();
```

I expected the full `SetChange` and all `PropertyChange`s for each changed `ChildEntity` in `ParentEnitity.childEntities` collecitons It works for non-audit diff with `javers.compare`.

Yes, I can further find the changes of the `ChildEntity` by the same commit, something like:
```java
javers
    .findChanges(
        QueryBuilder
            .byClass(ChildEntity.class)
            .withCommitId(new BigDecimal("4.00"))
            .build()
    )
    .groupByCommit()
```
, but this would be very complex since I have a complex objects hierarchy and doing this manually for each entity class would be cumbersome as well.

Finding all changes by `commitId` also looks not bad, but still does not contain the `SetChange` values for the collection fields:
```java
javers
    .findChanges(
        QueryBuilder
            .anyDomainObject()
            .withCommitId(new BigDecimal("4.00"))
            .build()
     )
     .groupByObject()
```

❗ Notably, for the first (`INITIAL`) commit, the `Changes` contains at least the `SetChange` for all entity collection fields.

❓ Is there a possibility to receive all entity tree changes in one commit with a single query? Or what is the "expected" way of handling `@OneToMany` in JaVers audit?

For `javers.compare` queries, the whole object tree comparison it's working nicely with annotating the `ChildEntity` either with javers `@Entity` or `@Id` annotations and with `withListCompareAlgorithm(ListCompareAlgorithm.AS_SET)` when creating `Javers`. Probably something similar should be used for the audit as well?

**Comment by dmitry-weirdo** [comment]
❗ **UPDATE:** Sorry, I probably was looking at the wrong commit. `SetChange` + `ValueAdded` is actually working for the parent entity collection field!

And I can actually further find all the changes with the same commit for all parent and child items hierarchy by the commit id:
```java
        javers
            .findChanges(
                QueryBuilder
                    .anyDomainObject()
                    .withCommitId(new BigDecimal("6.00")) // commitId where the parent entity was changed
                    .build()
            )
            .groupByObject()
```

❓  The question is now: can I get all these changes in one query. Something like `withChildValueObjects()`, but for the complete entity hierarchy?

Something like:

```java
        QueryBuilder builder = QueryBuilder
            .byInstanceId(parentId, ParentEntity.class) // only find changes within one instance of the root entity
            .withAllChildObjectChanges() // some magic method, does it exist?
        ;

        javers
            .findChanges(builder.build())
            .groupByCommit()
            .groupByObject() // is it possible to group by commit and within it by object?
```

**UPDATE 2:**
I actually found a way that performs a similar functionality:

**1)** Find shadows with a great `withScopeCommitDeep` parameter:
```java
        final List<Shadow<Object>> shadows = javers.findShadows(
            QueryBuilder
                .byInstanceId(parentEntityId, ParentEntity.class)
                .withScopeCommitDeep() // loads all child entities hierarchy, yes!!!
                .build()
        );
```

**2)** Compare each  `i + 1`-th and `i`-th pair of shadows:

```java
   javers.compare(shadows.get(1).get(), shadows.get(0).get()); // for each (i + 1)-th and i-th shadows
```

This is probably the expected way. The only thing I have in doubt is that this is exactly the same as what I did with Envers: `repository.findRevisions(parentEntityId)` + compare `Revisions` with `javers.compare`. So still not 146% sure which benefits JaVers will bring (probably it is the DB initialization on the existing database, as questioned in https://github.com/javers/javers/discussions/1116)...

**Reply by bartoszwalacik** [reply]
I can't help you with Envers. Please try to distill your question.
